File: apps/proxy/live_proxy/tests/harness/upstream.py
"""An in-process HTTP upstream for backend tests.

Not a reuse of the e2e-upstream Docker image on purpose: the backend test runner
is itself a container, and a sibling container per test is a poor trade for
something that has to start in milliseconds. This serves one looping TS payload
on one path, with the eight live-TS faults harness.faults ports.

Threading, not gevent: manage.py test does not monkey-patch (the patch is
applied only by dispatcharr/gevent_patch.py, imported from the uWSGI ini files),
so ThreadingHTTPServer's real OS threads are the right primitive and are the
same primitive the relay's own workers use under test.
"""

import http.server
import threading
import time

from .asset import TS_PACKET_SIZE, synthetic_ts
from .faults import FaultStore

_HTML_ERROR_PAGE = (
    b"<html><head><title>502 Bad Gateway</title></head>"
    b"<body><h1>502 Bad Gateway</h1></body></html>"
)

# What "rate 1.0" means, in bytes per second: 2 Mbit/s, the bitrate
# e2e-upstream/scripts/make-asset.sh builds its own asset at. slow-trickle's
# rate multiplies this.
NOMINAL_BYTE_RATE = 2_000_000 // 8

_WRITE_CHUNK = TS_PACKET_SIZE * 50  # 9,400 bytes


class _Handler(http.server.BaseHTTPRequestHandler):
    protocol_version = "HTTP/1.1"

    # Silence BaseHTTPRequestHandler's stderr access log; a test that wants to
    # know what was requested reads server.request_count.
    def log_message(self, *args):
        pass

    def do_GET(self):  # noqa: N802 - the stdlib's name
        upstream = self.server.upstream
        faults = upstream.faults
        with upstream._lock:
            upstream.request_count += 1
            hop = upstream._redirect_hops

        if faults.is_active("not-found"):
            self._send_short(404, b'{"error": "fault: not-found"}', "application/json")
            return
        if faults.is_active("auth-failure"):
            self._send_short(401, b'{"error": "fault: auth-failure"}', "application/json")
            return
        if faults.is_active("connection-limit"):
            self._send_short(429, b'{"error": "fault: connection-limit"}', "application/json")
            return

        if faults.is_active("redirect-chain"):
            depth = faults.config_of("redirect-chain")["depth"]
            if hop < depth:
                with upstream._lock:
                    upstream._redirect_hops += 1
                self.send_response(302)
                self.send_header("Location", upstream.url)
                self.send_header("Content-Length", "0")
                self.end_headers()
                return
            # Chain exhausted: this request serves the stream, and the next
            # connection starts a fresh chain.
            with upstream._lock:
                upstream._redirect_hops = 0

        if faults.is_active("non-ts-bytes"):
            self._send_short(200, _HTML_ERROR_PAGE, "text/html")
            return

        self.send_response(200)
        self.send_header("Content-Type", "video/mp2t")
        self.send_header("Connection", "close")
        self.end_headers()

        if faults.is_active("dead-air"):
            # Headers sent, nothing else, socket held open -- exactly what a
            # provider that stops producing looks like from the relay's side.
            deadline = time.monotonic() + upstream.dead_air_hold_seconds
            while time.monotonic() < deadline and not upstream._stopping.is_set():
                time.sleep(0.05)
            return

        self._stream(upstream)

    def _send_short(self, status, body, content_type):
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Connection", "close")
        self.end_headers()
        self.wfile.write(body)

    def _stream(self, upstream):
        payload = upstream.payload
        faults = upstream.faults

        rate = None
        if faults.is_active("slow-trickle"):
            rate = faults.config_of("slow-trickle")["rate"] * NOMINAL_BYTE_RATE
        elif upstream.rate is not None:
            rate = upstream.rate * NOMINAL_BYTE_RATE

        stop_after = None
        clean = True
        if faults.is_active("disconnect"):
            config = faults.config_of("disconnect")
            stop_after = config["after_bytes"]
            clean = config["clean"]

        sent = 0
        at = 0
        started = time.monotonic()
        try:
            while not upstream._stopping.is_set():
                if stop_after is not None and sent >= stop_after:
                    if not clean:
                        # Abrupt: drop the socket without a clean shutdown, so
                        # the reader sees a connection reset rather than EOF.
                        self.close_connection = True
                        try:
                            self.connection.close()
                        except OSError:
                            pass
                    return
                chunk = _WRITE_CHUNK
                if stop_after is not None:
                    chunk = min(chunk, stop_after - sent)
                piece = bytearray()
                while len(piece) < chunk:
                    take = min(chunk - len(piece), len(payload) - at)
                    piece += payload[at : at + take]
                    at = (at + take) % len(payload)
                self.wfile.write(bytes(piece))
                sent += len(piece)
                if rate is not None:
                    due = started + sent / rate
                    now = time.monotonic()
                    if due > now:
                        time.sleep(min(due - now, 0.25))
        except (BrokenPipeError, ConnectionResetError, OSError):
            # The relay closed its side -- an ordinary end to a tune.
            return


class _Server(http.server.ThreadingHTTPServer):
    daemon_threads = True
    allow_reuse_address = True

    def handle_error(self, request, client_address):
        # A tune ending, or the `disconnect` fault firing, closes a socket
        # mid-response. socketserver's default handle_error prints a full
        # traceback to stderr for each one, which would bury every real test
        # failure in noise. Swallowed here deliberately: nothing about this
        # server's own errors is under test, and a test that cares asserts on
        # what the client received.
        pass


class FakeUpstream:
    """A looping TS upstream on 127.0.0.1, with injectable faults.

    `payload` defaults to 512 synthetic packets (~96 KB). `rate` paces every
    response at that multiple of NOMINAL_BYTE_RATE even with no slow-trickle
    fault armed, and **defaults to 1.0 rather than None on purpose**: at
    loopback speed this server pushes about 34 MB/s into the ring buffer, which
    was measured filling Redis to 2.17 GB across 8,863 chunk keys in ~50
    seconds of a single held-open tune. Redis runs with `maxmemory 0` and DB 0
    is shared with the Celery broker and the Django cache, so an unpaced tune
    held open for a dead-air cycle or a buffering window -- which is precisely
    what 2a-3 … 2a-6 will do -- takes the whole test process down with it.
    `rate=None` restores the unpaced behaviour and is a deliberate,
    short-tune-only choice.
    """

    dead_air_hold_seconds = 30.0

    def __init__(self, payload: bytes | None = None, rate: float | None = 1.0) -> None:
        self.payload = payload if payload is not None else synthetic_ts(packets=512)
        if not self.payload:
            raise ValueError("payload must not be empty")
        self.rate = rate
        self.faults = FaultStore()
        self.request_count = 0
        self._redirect_hops = 0
        self._lock = threading.Lock()
        self._stopping = threading.Event()
        self._server: _Server | None = None
        self._thread: threading.Thread | None = None
        self.url = ""

    def start(self) -> "FakeUpstream":
        if self._server is not None:
            raise RuntimeError("already started")
        self._server = _Server(("127.0.0.1", 0), _Handler)
        self._server.upstream = self
        host, port = self._server.server_address[:2]
        self.url = f"http://{host}:{port}/live.ts"
        # poll_interval, not the 0.5s default: serve_forever() sleeps that long
        # between checks of its shutdown flag and shutdown() waits one out, so
        # the default costs every single upstream test half a second of pure
        # teardown -- measured at ~0.5s per test across the whole label.
        self._thread = threading.Thread(
            target=self._server.serve_forever, args=(0.01,), name="fake-upstream", daemon=True
        )
        self._thread.start()
        return self

    def stop(self) -> None:
        self._stopping.set()
        if self._server is not None:
            self._server.shutdown()
            self._server.server_close()
            self._server = None
        if self._thread is not None:
            self._thread.join(timeout=5)
            self._thread = None

    def __enter__(self) -> "FakeUpstream":
        return self.start()

    def __exit__(self, *exc) -> None:
        self.stop()

File: relay/internal/relaytest/upstream.go
package relaytest

import (
	"errors"
	"net/http"
	"net/http/httptest"
	"sync"
	"time"
)

// WriteChunk is how much the upstream writes to the wire at a time: 50 TS
// packets, apps/proxy/live_proxy/tests/harness/upstream.py:31's
// _WRITE_CHUNK. Pinned by TestNominalByteRateAndWriteChunkMatchThePythonHarness.
const WriteChunk = PacketSize * 50

// Config shapes one Upstream. The zero value is a valid unpaced upstream
// serving 512 synthetic packets on a loop, which is what most tests want.
type Config struct {
	// Payload is the asset to loop. Nil means SyntheticTS(512, 0x100).
	Payload []byte

	// Status is the response status. Zero means 200; any other value is sent
	// with a short JSON body and no stream.
	Status int

	// Rate paces the response at Rate * NominalByteRate bytes per second.
	// Zero means unpaced.
	//
	// UNPACED BY DEFAULT, and that is a DELIBERATE divergence from
	// harness/upstream.py, whose FakeUpstream defaults to 1.0. Its docstring
	// gives the reason: unpaced, it pushed ~34 MB/s into Redis and filled DB
	// 0 -- shared with the Celery broker and the Django cache -- to 2.17 GB in
	// 50 seconds, taking the test process down. Spec D2 deletes that failure
	// mode: the Go ring is bounded at MaxBytesPerChannel per channel by
	// construction, so an unpaced upstream costs a bounded 73 MiB and a much
	// faster test. Set Rate when the test's subject is throughput.
	Rate float64

	// StopAfterBytes ends the response after this many bytes. Zero means the
	// loop runs until the client goes away.
	StopAfterBytes int

	// Abrupt makes StopAfterBytes abort the response rather than end it
	// cleanly, so the reader sees a broken connection instead of EOF.
	Abrupt bool

	// DeadAir sends the 200 and the headers, then nothing at all for this
	// long -- exactly what a provider that stops producing looks like from
	// the relay's side. 2c-5's dead-air trigger (parity-matrix row 2) acts
	// on it.
	DeadAir time.Duration

	// DeadAirAfterBytes sends this many bytes and then nothing at all until
	// the client goes away: a provider that STOPPED rather than one that
	// never started, which is the shape that takes the relay past its
	// init grace period and onto CONNECTION_TIMEOUT (input/manager.py:
	// 1547-1551). Zero means no dead air. harness/standin.py's
	// --dead-air-after-bytes, on the upstream rather than the child.
	DeadAirAfterBytes int
}

// Upstream is a looping TS provider on 127.0.0.1.
type Upstream struct {
	server *httptest.Server

	mu       sync.Mutex
	requests int
	headers  []http.Header
	methods  []string
}

// NewUpstream starts an upstream. Register its Close with t.Cleanup.
func NewUpstream(cfg Config) *Upstream {
	payload := cfg.Payload
	if payload == nil {
		payload = SyntheticTS(512, 0x100)
	}
	if len(payload) == 0 {
		panic("relaytest: payload must not be empty")
	}

	u := &Upstream{}
	u.server = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		u.mu.Lock()
		u.requests++
		u.headers = append(u.headers, r.Header.Clone())
		u.methods = append(u.methods, r.Method)
		u.mu.Unlock()
		u.serve(w, r, cfg, payload)
	}))
	return u
}

// URL is the upstream's stream URL.
func (u *Upstream) URL() string { return u.server.URL + "/live.ts" }

// Requests is how many HTTP requests the upstream has answered. The count a
// "three clients share exactly one upstream connection" assertion reads.
func (u *Upstream) Requests() int {
	u.mu.Lock()
	defer u.mu.Unlock()
	return u.requests
}

// Headers is the header set of each request the upstream answered, in order.
// Cloned at receipt, so a test reading them races nothing.
func (u *Upstream) Headers() []http.Header {
	u.mu.Lock()
	defer u.mu.Unlock()
	return append([]http.Header(nil), u.headers...)
}

// Methods is the HTTP method of each request the upstream answered, in
// order -- how a test tells a HEAD probe from a GET that would have streamed.
func (u *Upstream) Methods() []string {
	u.mu.Lock()
	defer u.mu.Unlock()
	return append([]string(nil), u.methods...)
}

// Close stops the server and waits for its handlers.
func (u *Upstream) Close() { u.server.Close() }

func (u *Upstream) serve(w http.ResponseWriter, r *http.Request, cfg Config, payload []byte) {
	if cfg.Status != 0 && cfg.Status != http.StatusOK {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(cfg.Status)
		_, _ = w.Write([]byte(`{"error":"relaytest: configured status"}`))
		return
	}

	w.Header().Set("Content-Type", "video/mp2t")
	w.WriteHeader(http.StatusOK)
	rc := http.NewResponseController(w)
	if err := rc.Flush(); err != nil {
		return
	}

	if cfg.DeadAir > 0 {
		select {
		case <-r.Context().Done():
		case <-time.After(cfg.DeadAir):
		}
		return
	}

	var rate float64
	if cfg.Rate > 0 {
		rate = cfg.Rate * NominalByteRate
	}

	sent := 0
	at := 0
	started := time.Now()
	for {
		if cfg.StopAfterBytes > 0 && sent >= cfg.StopAfterBytes {
			if cfg.Abrupt {
				// The stdlib's own way to end a response without a clean
				// terminator. It also suppresses the server's stack trace,
				// which an ordinary panic would print over every test.
				panic(http.ErrAbortHandler)
			}
			return
		}
		if r.Context().Err() != nil {
			return
		}
		if cfg.DeadAirAfterBytes > 0 && sent >= cfg.DeadAirAfterBytes {
			// Connected, silent, and still here: the relay's watchdog is
			// what ends this, by hanging up.
			<-r.Context().Done()
			return
		}

		want := WriteChunk
		if cfg.StopAfterBytes > 0 && cfg.StopAfterBytes-sent < want {
			want = cfg.StopAfterBytes - sent
		}
		if cfg.DeadAirAfterBytes > 0 && cfg.DeadAirAfterBytes-sent < want {
			want = cfg.DeadAirAfterBytes - sent
		}
		piece := make([]byte, 0, want)
		for len(piece) < want {
			take := min(want-len(piece), len(payload)-at)
			piece = append(piece, payload[at:at+take]...)
			at = (at + take) % len(payload)
		}

		if _, err := w.Write(piece); err != nil {
			// The relay closed its side -- an ordinary end to a tune.
			return
		}
		if err := rc.Flush(); err != nil && !errors.Is(err, http.ErrNotSupported) {
			return
		}
		sent += len(piece)

		if rate > 0 {
			due := started.Add(time.Duration(float64(sent) / rate * float64(time.Second)))
			// Sleep UNTIL DUE, in steps of at most 250 ms so a client that
			// has gone is noticed within a step rather than after the whole
			// wait. An earlier form slept min(wait, 250ms) ONCE, capping the
			// total wait: any rate below ~37,600 B/s was delivered faster
			// than configured -- Rate 0.05 fixtures about three times too
			// fast, row 4's quarter-rate upstream looping every 13.5 s
			// instead of 32 s (issue #300, found by the 2c-4 review).
			for {
				wait := time.Until(due)
				if wait <= 0 || r.Context().Err() != nil {
					break
				}
				time.Sleep(min(wait, 250*time.Millisecond))
			}
		}
	}
}
